<!-- localization.launch -->
<!--
    This file launches a localization node that
    fuses all movement data (wheel odometry, IMU,
    GPS) into one filtered model for increased
    accuracy of all the constituent parts.
-->
<launch>

    <include file="$(find igvc_navigation)/launch/swerve_wheel_odometry.launch" />

    <node pkg="robot_localization" type="navsat_transform_node" name="navsat_transform_node" respawn="true">

      <param name="frequency" value="30"/>

      <!-- delay time could be useful but defaults to 0 -->
      <param name="delay" value="0"/>

      <!-- does ardupilot account for magnetic declination? -->
      <param name="magnetic_declination_radians" value="0"/>

      <!-- this assumes we are in a planar environment -->
      <param name="two_d_mode" value="false"/>

      <!-- Most IMU's report 0 as facing north. REP103 expects 0 when facing east. This param corrects for that -->
      <param name="yaw_offset" value="0"/>

      <!-- sets altitude to 0 -->
      <param name="zero_altitude" value="true"/>

      <!-- broadcast the transform between the UTM grid and the frame of the input odometry data -->
      <param name="broadcast_cartesian_transform" value="true"/>

      <!-- publishes the lat and long of the robot position -->
      <param name="publish_filtered_gps" value="true"/>

      <!-- if true uses yaw angle from the input odom source -->
      <param name="use_odometry_yaw" value="false"/>
    
      <param name="wait_for_datum" value="false"/>

      <!-- input remapping -->
      <remap from="/imu/data" to="/magnetometer"/>
      <remap from="/gps/fix" to="/fix"/>

      <!-- can do output remapping also -->
    </node>

    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" output="screen" required="true">
        <rosparam command="load" file="$(find igvc_navigation)/config/ekf_localization_node_params.yaml"/>
    </node>

    <node pkg="igvc_utils" type="quaternion_to_rpy" name="odometry_filtered_to_rpy"  output="screen" required="true">
        <param name="topics/quaternion" value="/odometry/filtered"/>
        <param name="topics/rpy" value="/odometry/rpy"/>
        <param name="message_type" value="odometry"/>
    </node>

</launch>
